enhancements/0002-20221026-experimental-rest-transport-layer.md

```diff
@@ -0,0 +1,156 @@
+- *Client authentication.* There should be a forward-compatible path for adding client authentication, but client authentication does not need to be implemented in the first version.
+- *Compatibility with `start_server`.*
+- *Build a "RESTful" API.* The goal is to have a REST-inspired API, but not to build a fully idiomatic RESTful API.
+
+## Proposal
+
+Introduce a request/response REST transport layer. The REST transport layer would consist of a REST API server (integrated with the Flower server) and a REST client.
+
+## Design Details
+
+### Definition(s)
+
+Task: User-defined specification of what a client has to do, with the expectation that the client returns a result.
+
+### Protocol
+
+General idea:
+- Client always initiates conversation
+- Server responds in one of two ways:
+  - Reconnect at a later time.
+  - Execute a task and send the result.
+- Server has a (mild) form of control over the conversation by suggesting to the client when it should connect again.
+
+Example:
+- Client connects to the server: "I'm available"
+- Server doesn't have any tasks: "reconnect in 5min"
+- Client sleeps for 5min, then connects again
+- Server say: here's a TASK and a TOKEN, work on it and send me your RESULT
+- Client receives the TASK, executes it locally, and returns the RESULT
+- Server says thanks, please reconnect in 10min
+- [the process repeats]
+
+### Server-side REST API
+
+Situation: a `Task`/`TaskAssignment` scheduled for `N` anonymous clients
+
+**GET /tasks**
+Request body: `None`
+Response body: `GetTasksResponse` containing either `Task` + `token` or `Reconnect`
+
+Implementation:
+1. Is there any task scheduled? Yes, one for `N` anonymous clients
+2. Generate the first `token`, store it with the `Task`/`TaskAssignment` in the `State`
+3. Return `Task` + `token` to the client in the `GetTasksResponse`
+
+**POST /results**
+
+Request body: `SubmitResultRequest` containing `Result` + `token`
+Response body: `None`
+
+Implementation:
+1. Check if the `token` is known
+2. Save `Result` if true, else return error
+
+### Message serialization
+
+TODO:
+- Continue to use ProtoBuf for serialization.
+- Send/receive serialized ProtoBuf messages.
+- `Content-Type` header must be set to `application/protobuf`.
+- Clients are expected to send an `Accept` header indicating that they accept `application/protobuf` responses. Clients must check the response `Content-Type` to ensure forward-compatibility.
+
+### Message types
+
+Messages types should stay identical to the existing ones, particularly the ones used in the Driver API:
+
+- `Task` wrapping a `ServerMessage` with its nested message types (`FitIns`, `EvaluateIns`, ...)
+- `Result` wrapping a `ClientMessage` with its nested message types (`FitRes`, `EvaluateRes`, ...) 
+
+### Security considerations
+
+**HTTPS-only:** Clients are expected to connect to the server only via HTTPS, never via plain HTTP.
+
+**Poisoning attacks:** [TODO, which mechanisms do we want to support/enforce here?]
```

Comment:
would be very interesting to explore https://privacypass.github.io/ as a solution